R/convert.R

```diff
@@ -333,16 +333,116 @@ bblDatToNonmem <- function(model, data, table=nlmixr2est::tableControl(), env=NU
+  betterOldCols <- c("id", "cmt", "dvid")
+
+  oldData <- data
+  oldData$nlmixrRowNums <- seq_len(nrow(oldData))
+  oldStdNames <- na.omit(getStandardColNames(oldData))
+  newStdNames <- na.omit(getStandardColNames(newData))
+
+  # Standardized column names to drop from the old and new data
+  prepDropFromOld <- betterNewCols[betterNewCols %in% names(newStdNames)]
+  prepDropFromNew <- betterOldCols[betterOldCols %in% names(oldStdNames)]
+  # Actual column names to drop from the old and new data
+  dropFromOld <- oldStdNames[prepDropFromOld]
+  dropFromNew <- newStdNames[prepDropFromNew]
+
+  # Prepare for merging and merge
+  oldDataPrep <- oldData[, setdiff(names(oldData), dropFromOld), drop=FALSE]
+  newDataPrep <- newData[, setdiff(names(newData), dropFromNew), drop=FALSE]
+  stopifnot(intersect(names(oldDataPrep), names(newDataPrep)) == "nlmixrRowNums")
+  mergedData <- merge(oldDataPrep, newDataPrep, by = "nlmixrRowNums")
+  cleanStdNames <- getStandardColNames(mergedData)
+
+  # Extract out the observation and dosing data
+  obsData <- mergedData[mergedData[[cleanStdNames[["evid"]]]] == 0, ]
+  if (nrow(obsData) < 1) {
+    cli::cli_abort("no observation rows (EVID = 0) detected")
+  }
+  doseData <- mergedData[mergedData[[cleanStdNames[["evid"]]]] %in% c(1, 4), ]
+  if (nrow(doseData) < 1) {
+    cli::cli_abort("no dosing rows (EVID = 1 or 4) detected")
+  }
+  obsCmt <- unique(obsData[[cleanStdNames[["cmt"]]]])
+  doseCmt <- unique(doseData[[cleanStdNames[["cmt"]]]])
+  stopifnot(length(obsCmt) == 1)
+  stopifnot(length(doseCmt) == 1)
+
+  # Drop subjects in only one dataset
+  dropDoseCmt <- !(doseData[[cleanStdNames["id"]]] %in% unique(obsData[[cleanStdNames["id"]]]))
+  dropObsCmt <- !(obsData[[cleanStdNames["id"]]] %in% unique(doseData[[cleanStdNames["id"]]]))
+  if (any(dropDoseCmt)) {
+    cli::cli_alert_info(paste("Dropping", sum(dropDoseCmt), "dosing rows with no observations for PKNCA estimation"))
+    doseCmt <- doseCmt[!dropDoseCmt, ]
+  }
+  if (any(dropObsCmt)) {
+    cli::cli_alert_info(paste("Dropping", sum(dropObsCmt), "observation rows with no doses for PKNCA estimation"))
+    obsCmt <- obsCmt[!dropObsCmt, ]
+  }
+
+  if (nrow(obsData) < 1) {
+    cli::cli_abort("No observation rows after filtering for analysis")
+  } else if (nrow(doseData) < 1) {
+    cli::cli_abort("No dose rows after filtering for analysis")
+  }
+  list(
+    obs=obsData,
+    dose=doseData
+  )
+}
+
+#' Determine standardized rxode2 column names from data
+#'
+#' @param data A data.frame as the source for column names
+#' @return A named character vector where the names are the standardized names
+#'   and the values are either the name of the column from the data or \code{NA}
+#'   if the column is not present in the data.
+#' @examples
+#' getStandardColNames(data.frame(ID=1, DV=2, Time=3, CmT=4))
+#' @export
+getStandardColNames <- function(data) {
```

Comment:
@mattfidler, This is the function I'm thinking of for #52.  I'll give more description there.